src/rgen/Microsoft.Macios.Generator/TabbedStringBuilder.cs

```diff
@@ -0,0 +1,228 @@
+/// using (var getConstantBlock = classBlock.CreateBlock (isBlock: true)) {
+///    // write the contents of the method here.
+/// }
+/// </example>
+/// </summary>
+class TabbedStringBuilder : IDisposable {
+	readonly StringBuilder sb;
+	readonly uint tabCount;
+	readonly bool isBlock;
+	bool disposed;
+
+	/// <summary>
+	/// Created a new tabbed string builder that will use the given sb to write code.
+	/// </summary>
+	/// <param name="builder">The string builder to be used to write code.</param>
+	/// <param name="currentCount">The original tab size.</param>
+	/// <param name="block">States if we are creating a {} block.</param>
+	public TabbedStringBuilder (StringBuilder builder, uint currentCount = 0, bool block = false)
+	{
+		sb = builder;
+		isBlock = block;
+		if (isBlock) {
+			// increase by 1 because we are in a block
+			tabCount = currentCount;
+			WriteTabs ().Append ('{').AppendLine ();
+			tabCount++;
+		} else {
+			tabCount = currentCount;
+		}
+	}
+
+	StringBuilder WriteTabs () => sb.Append ('\t', (int) tabCount);
+
+	/// <summary>
+	/// Append a new empty line to the string builder using the correct tab size.
+	/// </summary>
+	/// <returns>The current tabbed string builder.</returns>
+	public TabbedStringBuilder AppendLine ()
+	{
+		sb.AppendLine ();
+		return this;
+	}
+
+	/// <summary>
+	/// Append conteng, but do not add a \n
+	/// </summary>
+	/// <param name="line">The content to append.</param>
+	/// <returns>The current builder.</returns>
+	public TabbedStringBuilder Append (string line)
+	{
+		if (string.IsNullOrWhiteSpace (line)) {
+			sb.Append (line);
+		} else {
+			WriteTabs ().Append (line);
+		}
+		return this;
+	}
+
+	/// <summary>
+	/// Append conteng, but do not add a \n
+	/// </summary>
+	/// <param name="span">The content to append.</param>
+	/// <returns>The current builder.</returns>
+	public TabbedStringBuilder Append (ReadOnlySpan<char> span)
+	{
+		if (span.IsWhiteSpace ()) {
+			sb.Append (span);
+		} else {
+			WriteTabs ().Append (span);
+		}
+
+		return this;
+	}
+
+	/// <summary>
+	/// Append a new tabbed line.
+	/// </summary>
+	/// <param name="line">The line to append.</param>
+	/// <returns>The current builder.</returns>
+	public TabbedStringBuilder AppendLine (string line)
+	{
+		if (string.IsNullOrWhiteSpace (line)) {
+			sb.AppendLine (line);
+		} else {
+			WriteTabs ().AppendLine (line);
+		}
+		return this;
+	}
+
+	/// <summary>
+	/// Append a new tabbed lien from the span.
+	/// </summary>
+	/// <param name="span">The line to append.</param>
+	/// <returns>The current builder.</returns>
+	public TabbedStringBuilder AppendLine (ReadOnlySpan<char> span)
+	{
+		if (span.IsWhiteSpace ()) {
+			sb.Append (span).AppendLine ();
+		} else {
+			WriteTabs ().Append (span).AppendLine ();
+		}
+
+		return this;
+	}
+
+	public TabbedStringBuilder Append (ref DefaultInterpolatedStringHandler handler)
+	{
+		WriteTabs ().Append (handler.ToStringAndClear ());
```

Comment:
This doesn't work? The whole point of this overload is to not have to create a string instance.

```suggestion
		WriteTabs ().Append (ref handler);
```